### DISCUSSION TITLE
How to release focus trap of a Menu (and keep it on the screen)?

### DISCUSSION THREAD

**Discussion by rfgamaral** [ariakit/ariakit#3101] [Questions] [answered]
Hello there 👋

In this app that I'm working on, I'm trying to refactor a legacy menu component to use the Menu component from `ariakit`, but this is proving to be a challenge given that certain menu items open custom popovers.

At the moment, both the legacy menu component and the custom popovers are built with `react-popper` and `react-focus-lock`, and everything works as expected. I have pretty much finished a refactor for the menu component using `ariakit` - it was a relatively straightforward process - but opening a custom popover when certain items are clicked, and **shifting focus** to the _new_ popover, is what I'm struggling with.

> [!NOTE]  
> The end goal is to eventually refactor these custom popovers, possibly with `Popover` from `ariakit`, but this is not something that I can prioritize at the moment. So, hopefully, someone can help me figure out a workaround to my problem.

It's my understanding that `ariakit` has its own "focus trap" mechanism implemented, and from what I could observe, this is competing with the "focus lock" that we have implemented in our popovers implemented with `react-popper`. From `react-focus-lock` [documentation](https://github.com/theKashey/react-focus-lock#focus-fighting):

> Two different focus-lock-managers or even different version of a single one, active simultaneously will FIGHT!
> 
> Focus-lock will surrender, as long any other focus management library will not.

Which is what I believe it's the source of the issue I'm having.

The second popover, the custom one that opens after clicking a menu item inside the first popover (the `ariakit` menu), surrenders the focus lock to the menu, making it impossible to interact with, despite the fact that it's being rendered on top of the menu.

So, my question is, how can I surrender the focus lock **from the `ariakit` menu** when opening any of these custom popovers? Ideally, the menu doesn't hide when clicking the menu item, so that the custom popover can be anchored to the menu item that triggered it.

**Comment by diegohaz** [comment]
Hi @rfgamaral! [Ariakit **does not** trap focus](https://ariakit.org/examples/dialog-radix#focus-trap-elements). If you can provide a reproduction on StackBlitz or CodeSandbox using react-popper, I can take a look and give better instructions.

**Reply by rfgamaral** [reply]
> [Ariakit does not trap focus](https://ariakit.org/examples/dialog-radix#focus-trap-elements).

I forgot to mention that, unfortunately, we are still using `"ariakit": "2.0.0-next.43"`. Could it be that that version had a focus trap mechanism? I do see `data-focus-trap` attributes on some of the menu elements, and I believe I've read about it in some older issues.

Did it use to have focus trap, and has since been removed from the latest stable releases?

> If you can provide a reproduction on StackBlitz or CodeSandbox using react-popper, I can take a look and give better instructions.

I did my best to recreate a [minimal implementation in a CodeSandbox](https://codesandbox.io/p/sandbox/ariakit-menu-and-popper-cg6n3n), and I was able to reproduce the issues we're having. This uses the exact same versions that we currently use on our side, and if this is something that could be solved by upgrading to the latest `@ariakit/core` version, maybe it's something we'll need to prioritize. Please let me know. With that said, if you could still have a look at the reproduction with the older `ariakit` version and see if there's some workaround that we could apply in the meantime, I'd really appreciate that.

Now, in the current CodeSandbox, here's the result:

![PowerToys_29IWYodfu4](https://github.com/ariakit/ariakit/assets/96476/bd0478e2-6345-4253-a4c2-23c74bd4bd0d)

You can see that clicking the `Popper` menu item opens a `react-popper`-based component, but clicking on any element inside the popper, closes everything.

However, if we set `hideOnInteractOutside={false}` on the `Menu` component, clicking on elements inside the popper no longer closes the menu:

![PowerToys_1FWb7wcWng](https://github.com/ariakit/ariakit/assets/96476/c0097c61-97e8-4471-9219-9a7ddf90b9d2)

But we still can't interact with any of the elements inside the popper (clicking `Search` should show an alert window).

Short of refactoring a bunch of components on our side, and/or upgrading `ariakit`, I've exhausted all my ideas to work around this issue in a simpler and quicker manner. Upgrading components, and moving away from `react-popper` is something that we plan to do, but that's a very big effort, and I was looking for something quicker in the short term.

**Reply by rfgamaral** [reply]
@diegohaz I think I found the source of the issue for us, and it was a bit hidden in our menu abstraction implemented in `@doist/reactist`, where we set `modal: true` in the `Menu` component. Setting that to `false` seems to have fixed the issue. However, reading the documentation:

> Determines whether the dialog is modal. Modal dialogs have distinct states and behaviors:
>
> - The [`portal`](https://ariakit.org/reference/dialog#portal) and [`preventBodyScroll`](https://ariakit.org/reference/dialog#preventbodyscroll) props are set to `true`. They can still be manually set to `false`.
>
> - A visually hidden dismiss button will be rendered if the [`DialogDismiss`](https://ariakit.org/reference/dialog-dismiss) component hasn't been used. This allows screen reader users to close the dialog.
> - When the dialog is open, element tree outside it will be disabled.
> - When using the [`Heading`](https://ariakit.org/reference/heading) or [`DialogHeading`](https://ariakit.org/reference/dialog-heading) components within the dialog, their level will be reset so they start with `h1`.

If we focus on the third point - `When the dialog is open, element tree outside it will be disabled` - I believe that's the one that caused the original issue for us, right? I'm wondering, instead of setting `modal: false` (and maybe setting `portal: true`, and `preventBodyScroll: true`), is it possible to disable this behaviour where the element tree outside the menu is disabled (assuming that's the issue)? Any workaround just for that? Perhaps something that we could disable when a "popper" element is on top of the menu, and enable back when the "popper" element is gone?

**Reply by diegohaz** [reply, marked as answer]
Thank you for the comprehensive response and for reproducing the issue.

Indeed, the problem isn't present in the most recent versions of `@ariakit/react`, even when the `modal` prop is being used. The modal's behavior was revised in #2339 to accommodate third-party popups.

If a migration to `@ariakit/react` isn't feasible right now, I recommend turning off the `modal` prop while the popper element is visible:

```jsx
<MenuList portal modal={!isPopperShown} />
```

**Reply by rfgamaral** [reply]
Will give that a try tomorrow. Thanks for the reply. Will also explore the possibility of upgrading to the latest Ariakit version, but I think the effort required for that will be more than I can handle at the moment.

**Reply by rfgamaral** [reply]
Thank you for your assistance, @diegohaz 🙌
